src/proxy-animation.js

```diff
@@ -9,119 +9,350 @@ const nativeElementAnimate = window.Element.prototype.animate;
-  let sequence = ++details.sequence;
-  let promise = details.readyPromise = new Promise((resolve, reject) => {
-    // TODO: We should actually not apply the animation until this is
-    // resolved.
-    requestAnimationFrame(() => {
-      // If this promise was replaced, this animation was aborted.
-      if (details.readyPromise == promise)
-        details.readyPromise = null;
-      if (details.aborted.has(sequence)) {
-        details.aborted.delete(sequence);
-        // Reject with a non-visible AbortError.
-        reject(new DOMException("Animation aborted", "AbortError"));
-        return;
-      }
+  let nativeResolve = undefined;
+  let nativeReject = undefined;
+  let pendingTask = undefined;
+  let pendingTaskName = undefined;
+  let state = 'pending';
+  const p = new Promise((resolve, reject) => {
+    nativeResolve = resolve;
+    nativeReject = reject;
+  });
 
-      // See pending play and pause tasks in:
-      // https://www.w3.org/TR/web-animations-1/#playing-an-animation-section
-      // https://www.w3.org/TR/web-animations-1/#pausing-an-animation-section
+  p.resolve = () => {
+    state = 'resolved';
+    pendingTask = null;
+    pendingTaskName = null;
+    nativeResolve(details.proxy);
+  }
+  p.reject = () => {
+    state = 'rejected';
+    pendingTask = null;
+    pendingTaskName = null;
+    nativeReject(new DOMException("The user aborted a request", "AbortError"));
+  }
+  p.cancelTask = () => {
+    pendingTask = null;
+  }
+  p.queueTask = (task, name) => {
+    pendingTask = task;
+    pendingTaskName = name;
+  }
+  p.state = () => {
+    return state;
+  }
+  p.taskName = () => {
+    return pendingTaskName;
+  }
 
-      // TODO: Consider implementing commitPendingPlay and commitPendingPause
-      // functions with closer adherence to the spec.
+  const runOrRequeueTask = () => {
+    if (!pendingTask)
+      return;
 
-      // Apply pending playbackRate
-      if (details.pendingPlaybackRate !== null) {
-        const previousCurrentTime = details.animation.currentTime;
-        details.animation.playbackRate = details.pendingPlaybackRate;
-        const timelineTime = details.timeline.currentTime;
-        details.startTime = details.pendingPlaybackRate ?
-            timelineTime - previousCurrentTime / details.pendingPlaybackRate :
-            previousCurrentTime;
-        details.pendingPlaybackRate = null;
-      }
-      switch(details.playState) {
-        case 'paused':
-          details.startTime = null;
-          break;
+    // Switching from a scroll timeline to a document timeline while pending
+    // casues the timeline to become null.
+    if (details.timeline == null)
+      return;
 
-        case 'running':
-        case 'finished':
-          details.holdTime = null;
-      }
-      resolve();
-    });
+    if (details.timeline.currentTime !== null) {
+      pendingTask(details);
+      pendingTask = null;
+      return;
+    }
+    requestAnimationFrame(runOrRequeueTask);
```

Comment:
Running a requestAnimationFrame loop while waiting for the conditions with which to run the pending task introduces non-trivial cost to running this polyfill. We should be able to observe each of these conditions that may change (i.e. set timeline method, a call to tick the animation) and run the pending task when they do.

Reply:
done